pkg/storage/runtime.go

```diff
@@ -0,0 +1,418 @@
+	PodID   string `json:"pod-id"`   // Applicable to both PodSandboxes and Containers, mandatory
+	// The provided name and the ID of the image that was used to
+	// instantiate the container.
+	ImageName string `json:"image-name"` // Applicable to both PodSandboxes and Containers
+	ImageID   string `json:"image-id"`   // Applicable to both PodSandboxes and Containers
+	// The container's name, which for an infrastructure container is usually PodName + "-infra".
+	ContainerName string `json:"name"` // Applicable to both PodSandboxes and Containers, mandatory
+	// The name as originally specified in PodSandbox or Container CRI metadata.
+	MetadataName string `json:"metadata-name"`        // Applicable to both PodSandboxes and Containers, mandatory
+	UID          string `json:"uid,omitempty"`        // Only applicable to pods
+	Namespace    string `json:"namespace,omitempty"`  // Only applicable to pods
+	Attempt      uint32 `json:"attempt,omitempty"`    // Applicable to both PodSandboxes and Containers
+	CreatedAt    int64  `json:"created-at"`           // Applicable to both PodSandboxes and Containers
+	MountLabel   string `json:"mountlabel,omitempty"` // Applicable to both PodSandboxes and Containers
+}
+
+// SetMountLabel updates the mount label held by a RuntimeContainerMetadata
+// object.
+func (metadata *RuntimeContainerMetadata) SetMountLabel(mountLabel string) {
+	metadata.MountLabel = mountLabel
+}
+
+func (r *runtimeService) createContainerOrPodSandbox(metadata RuntimeContainerMetadata, containerID string) (ContainerInfo, error) {
+	if metadata.PodName == "" || metadata.PodID == "" {
+		return ContainerInfo{}, ErrInvalidPodName
+	}
+	if metadata.ContainerName == "" {
+		return ContainerInfo{}, ErrInvalidContainerName
+	}
+	if metadata.MetadataName == "" {
+		metadata.MetadataName = metadata.ContainerName
+	}
+	if metadata.CreatedAt == 0 {
+		metadata.CreatedAt = time.Now().Unix()
+	}
+
+	// Check if we have the specified image.
+	ref, err := istorage.Transport.ParseStoreReference(r.image.GetStore(), metadata.ImageName)
+	if err != nil {
+		// Maybe it's some other transport's copy of the image?
+		otherRef, err2 := transports.ParseImageName(metadata.ImageName)
+		if err2 == nil && otherRef.DockerReference() != nil {
+			ref, err = istorage.Transport.ParseStoreReference(r.image.GetStore(), otherRef.DockerReference().FullName())
+		}
+		if err != nil {
+			return ContainerInfo{}, err
+		}
+	}
+	img, err := istorage.Transport.GetStoreImage(r.image.GetStore(), ref)
+	if img == nil && err == storage.ErrImageUnknown && metadata.ImageID != "" {
+		ref, err = istorage.Transport.ParseStoreReference(r.image.GetStore(), "@"+metadata.ImageID)
+		if err != nil {
+			return ContainerInfo{}, err
+		}
+		img, err = istorage.Transport.GetStoreImage(r.image.GetStore(), ref)
+	}
+	if err != nil && err != storage.ErrImageUnknown {
+		return ContainerInfo{}, err
+	}
+
+	// Pull the image down if we don't already have it.
```

Comment:
Yeah, we should remove this but maybe as a separate PR. This will require us to change our tests to explicitly pull an image before starting pods and containers.